v0.30: swap indexes (#1992)
meilisearch/documentation: learn/core_concepts/indexes.md

You can update the typo tolerance settings using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update typo tolerance endpoint](/reference/api/settings.md#update-typo-tolerance-settings).

[Learn more about typo tolerance.](/learn/configuration/typo_tolerance.md)

## Swapping indexes

Suppose you want to deploy a new version of your `movies` index in production. This `movies` index is where your users are currently making search requests. The changes you made are in `movies_new`. To sync `movies_new` with Meilisearch in production, you will use index swapping.

This means that the schema, settings, and task history of  `movies` will be swapped with the schema, settings, and task history of `movies_new` **without any downtime for the search clients**. The task history of `enqueued` tasks is not modified.

> Once swapped, your users will still be making search requests to the `movies` index but it will contain the data of `movies_new`. You can delete `movies_new` after the swap or keep it in case something goes wrong and you want to swap back.

Reviewer: What about documents? They are not mentioned explicitly, but this phrase

> Once swapped, your users will still be making search requests to the `movies` index but **it will contain the data of `movies_new`**.

makes me think that swapping indexes also swaps the documents.

Swapping indexes is an atomic transaction: **either all indexes are successfully swapped, or none are**.

> You can perform multiple swap operations using the [swap indexes endpoint](/reference/api/indexes.md#swap-indexes).

Reviewer: The fact that you can perform multiple swap operations with a single API request is important but seems like a much better fit for the API reference than here. Since this is a core concepts document—not intended to be exhaustive—it's better to leave off practical details like this so that the key takeaways stand out more. Just my 2️⃣ 🪙 

```suggestion
For more information, see the [swap indexes endpoint](/reference/api/indexes.md#swap-indexes).
```

You can update the typo tolerance settings using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update typo tolerance endpoint](/reference/api/settings.md#update-typo-tolerance-settings).

[Learn more about typo tolerance.](/learn/configuration/typo_tolerance.md)

## Swapping indexes

> Suppose you want to update your `movies` index in production because you changed the schema or settings. The changes you made are in `movies_new`. To sync `movies_new` with Meilisearch in production, you will use index swapping. This means that the schema, settings, and task history of  `movies` will be swapped with the schema, settings, and task history of `movies_new` without any downtime for the search clients. The task history of `enqueued` tasks is not modified.

Reviewer 2: I would have put "Suppose you want to deploy a new index version in production **without any downtime**..." first

> Suppose you want to deploy a new version of your `movies` index in production without any downtime. The changes you made are in `movies_new`. To sync `movies_new` with Meilisearch in production, you will use index swapping. This means that the schema, settings, and task history of  `movies` will be swapped with the schema, settings, and task history of `movies_new`. The task history of `enqueued` tasks is not modified. Swapping indexes is an atomic transaction, **either all indexes are successfully swapped, or none are**.

Reviewer 2: - is one index in production and the other in dev?
    - `indexA` is served in production (users are making search request on it)
    - In the meantime you are building the future new index to be served in production. Let's call it `indexA_new`
- if my changes are in another dev, should I deploy them to production? Should I rename it?
     - When `indexA_new` is ready (everything has been indexed, the settings are fine etc) you want to deploy it to user.
    - You swap data (documents, settings, and tasks history) of `indexA` with `indexA_new`. Clients are still searching on `indexA` but `indexA` contains the data from `indexA_new`; `indexA_new` contains the `indexA` data. Clients experience a smooth experience. The deployment was made with 0 downtimes.
- what happens with the old index? Do I need to do something with it? Is it good practice to delete it afterwards?
     - This is to your appreciation. If everything is working well and that index won't be used anymore, it's perfectly okay to delete it.

> Suppose you want to deploy a new version of your `movies` index in production without any downtime. The changes you made are in `movies_new`. To sync `movies_new` with Meilisearch in production, you will use index swapping. This means that the schema, settings, and task history of  `movies` will be swapped with the schema, settings, and task history of `movies_new`. The task history of `enqueued` tasks is not modified. Swapping indexes is an atomic transaction, **either all indexes are successfully swapped, or none are**.

Reviewer 3: yes, that's very helpful, @gmourier, thanks!

I think at least some of this info should be on our docs. What we currently have is not bad, but I think it only really works if the reader knows what we're talking about. Adding some context, perhaps even in the form of a shortened version of the use-case @gmourier described, would make it much easier to grasp for first timers, IMHO.

> Suppose you want to deploy a new version of your `movies` index in production without any downtime. The changes you made are in `movies_new`. To sync `movies_new` with Meilisearch in production, you will use index swapping. This means that the schema, settings, and task history of  `movies` will be swapped with the schema, settings, and task history of `movies_new`. The task history of `enqueued` tasks is not modified. Swapping indexes is an atomic transaction, **either all indexes are successfully swapped, or none are**.

Author: Thanks @gmourier! 🙏

> You can perform multiple swap operations using the [swap indexes endpoint](/reference/api/indexes.md#swap-indexes).

Reviewer 2: Should we mention that these are atomic operations?
